language/move-compiler/src/parser/syntax.rs

```diff
@@ -1696,7 +1696,16 @@ fn parse_type(context: &mut Context) -> Result<Type, Box<Diagnostic>> {
         }
         Tok::Pipe => {
             let args = parse_comma_list(context, Tok::Pipe, Tok::Pipe, parse_type, "a type")?;
-            let result = parse_type(context)?;
+            let result = if is_start_of_type(context) {
```

Comment:
One possibility of removing this `is_start_of_type` is to call the `parse_type(context)` still and one of the following cases should happen:
- `parse_type` returns `Ok(t)`, then we take `t` as the return type
- `parse_type` returns `Err(_)` **AND** cursor does not advance, then we take `Unit` as a return type
- `parse_type` returns `Err(_)` **AND** cursor does advance, then we just propagate the diagnostic back